src/include/daos_fs_sys.h

```diff
@@ -0,0 +1,581 @@
+	D_ASSERT(d_hash_rec_unlinked(&hdl->entry));
+	dfs_sys_hash_hdl_delete(hdl);
+}
+
+/**
+ * Operations for the hash table.
+ */
+static d_hash_table_ops_t dfs_sys_hash_hdl_ops = {
+	.hop_key_cmp	= dfs_sys_hash_key_cmp,
+	.hop_rec_addref = dfs_sys_hash_rec_addref,
+	.hop_rec_decref	= dfs_sys_hash_rec_decref,
+	.hop_rec_free	= dfs_sys_hash_rec_free
+};
+
+
+/**
+ * Mount a file system with dfs_mount and optionally initialize the hash.
+ */
+/* TODO to be more consistent with DFS, maybe this header */
+// dfs_sys_mount(daos_handle_t poh, daos_handle_t coh, int flags, dfs_sys_t **_dfs_sys)
+/* TODO maybe let a different flags be passed for dfs_sys hashing/locking/future */
+int
+dfs_sys_mount(dfs_sys_t *dfs_sys, daos_handle_t *poh, daos_handle_t *coh,
+	      int flags, bool use_hash)
+{
+	int rc;
+
+	rc = dfs_mount(*poh, *coh, flags, &dfs_sys->dfs);
+	if (rc) {
+	D_ERROR("dfs_mount() failed (%d)\n", rc);
+		D_GOTO(out, rc);
+	}
+
+	/** TODO make this thread safe */
+	/** TODO allow size to be passed in? Default size? */
+	if (use_hash) {
+		rc = d_hash_table_create(D_HASH_FT_NOLOCK, 16, NULL,
+					 &dfs_sys_hash_hdl_ops,
+					&dfs_sys->dfs_hash);
+		if (rc) {
+			/** TODO error handling */
+			D_GOTO(err_hash, rc);
+		}
+	}
+
+err_hash:
+	dfs_umount(dfs_sys->dfs);
+out:
+	return rc;
+}
+
+/**
+ * Unmount a file system with dfs_mount and destroy the hash.
+ */
+int
+dfs_sys_umount(dfs_sys_t *dfs_sys)
+{
+	int rc;
+
+	/** TODO check params to this function */
+	rc = d_hash_table_destroy(dfs_sys->dfs_hash, true);
+	/** TODO error handling */
+
+	rc = dfs_umount(dfs_sys->dfs);
+	/** TODO error handling */
+
+	return rc;
+}
+
+
+/**
+ * Try to get name from the hash.
+ * If not found, call dfs_lookup on name
+ * and store in the hash.
+ * Stores the hashed obj in _obj.
+ */
+static int
+dfs_sys_hash_lookup(const char *name, dfs_sys_t *dfs_sys, dfs_obj_t **_obj)
+{
+	struct dfs_sys_hash_hdl	*hdl;
+	d_list_t		*rlink;
+	size_t			name_len;
+	int			rc;
+
+	/** Make sure the hash is initialized */
+	/** TODO optionally allow no caching? */
+	if (dfs_sys->dfs_hash == NULL)
+		return EINVAL;
+
+	name_len = strnlen(name, PATH_MAX);
+	if (name_len > PATH_MAX-1)
+		return ENAMETOOLONG;
+
+	/** If cached, return it */
+	rlink = d_hash_rec_find(dfs_sys->dfs_hash, name, name_len);
+	if (rlink != NULL) {
+		hdl = dfs_sys_hash_hdl_obj(rlink);
+		D_GOTO(out, rc = 0);
+	}
+
+	/** Not cached, so add it */
+	hdl = dfs_sys_hash_hdl_new();
+	if (hdl == NULL)
+		return ENOMEM;
+
+	D_STRNDUP(hdl->name, name, PATH_MAX);
+	if (hdl->name == NULL)
+		return ENOMEM;
+
+	/** Lookup name in dfs */
+	rc = dfs_lookup(dfs_sys->dfs, name, O_RDWR, &hdl->oh, NULL, NULL);
+	if (rc) {
+		D_ERROR("dfs_lookup() %s failed (%d)\n", name, rc);
+		D_GOTO(err_hdl, rc);
+	}
+
+	/* Store this entry in the hash.
+	 * Since we have already called d_hash_rec_find,
+	 * pass exclusive=false to avoid another find being called */
+	rc = d_hash_rec_insert(dfs_sys->dfs_hash, hdl->name, name_len,
```

Comment:
Not thread safe which is probably OK, but it means this whole interface isn't now.

As d_hash_rec_insert() can only fail with exclusive = True then you could use assert to check rc here.